### PR TITLE
feat!: 0.13 upgrade

### LINKED ISSUE #27: Add support for Terraform 0.13 for_each
- [x] Add support for TF 0.13
- [x] Add example to support multiple SLOs with `for_each` module attributes
- [X] Update dev tools version to 0.13 and kitchen `verify_version`

### PR DESCRIPTION
**Pull request by ocervell**
Fix #27

### CODE REVIEW

**Comment by bharathkkb** [comment]
@ocervell I think we were planning on supporting 0.12 till end of Q4? Is this just being staged?
/cc @morgante

**Comment by morgante** [comment]
I don't mind if we start introducing 0.13 on this module since it's not one of the core infra modules.

**Comment by ocervell** [comment]
@morgante we might need to fix CI though:
```
Step #0 - "prepare": Error: Unsupported Terraform Core version
Step #0 - "prepare": 
Step #0 - "prepare":   on .terraform/modules/project/versions.tf line 18, in terraform:
Step #0 - "prepare":   18:   required_version = "~> 0.12.6"
Step #0 - "prepare": 
Step #0 - "prepare": Module module.project (from terraform-google-modules/project-factory/google)
Step #0 - "prepare": does not support Terraform version 0.13.0. To proceed, either choose another
Step #0 - "prepare": supported Terraform version or update this version constraint. Version
Step #0 - "prepare": constraints are normally set for good reason, so updating the constraint may
Step #0 - "prepare": lead to other errors or unexpected behavior.
```

**Comment by morgante** [comment]
@ocervell There is a 0.13 Terraform devtools image available (tag 0.13).

Also make sure you're using the latest project factory version.

**Comment by ocervell** [comment]
Still getting an issue with latest PF and Dev tools 0.13.0:
```
Step #1 - "create": -----> Creating <simple-example-default>...
Step #1 - "create":        
Step #1 - "create":        Your version of Terraform is out of date! The latest version
Step #1 - "create":        is 0.13.3. You can update by downloading from https://www.terraform.io/downloads.html
Step #1 - "create":        Terraform v0.13.0
Step #1 - "create": >>>>>> ------Exception-------
Step #1 - "create": >>>>>> Class: Kitchen::ActionFailed
Step #1 - "create": >>>>>> Message: 1 actions failed.
Step #1 - "create": >>>>>>     Create failed on instance <simple-example-default>.  Please see .kitchen/logs/simple-example-default.log for more details
```
Seems to be from Kitchen.

**Comment by morgante** [comment]
@ocervell Please add [this line](https://github.com/terraform-google-modules/terraform-google-cloud-operations/blob/master/kitchen.yml#L32) to the kitchen.yaml

**Comment by ocervell** [comment]
@morgante could you please review my last commit ?

I noticed Terraform 0.13 uses only the one folder to store module code for every child of the for_each loop in `.terraform/modules/<module_name>`, which means my GCF files were overwritten successively by the different invocations in the `for_each` loop. 

Had to create a separate directory to store the GCF code in so that they don't conflict

File: modules/slo/main.tf [hunk 1]
```diff
@@ -17,32 +17,45 @@
...
+  function_source_directory = var.function_source_directory != "" ? var.function_source_directory : "${path.module}/code"
+  function_target_directory = "${path.module}/code_${local.suffix}"
   requirements_txt = templatefile(
     "${path.module}/code/requirements.txt.tpl", {
       slo_generator_version = var.slo_generator_version
     }
   )
 }
 
+resource "null_resource" "copy_gcf_folder" {
```
Comment (morgante) [inline modules/slo/main.tf]: I'm not sure this is necessary? Last I checked the `local_file` resource autocreates parent directories.

**Review by morgante — CHANGES_REQUESTED** [review]

File: modules/slo/main.tf (hunk 1 above)
Comment (ocervell) [reply modules/slo/main.tf]: ah, indeed it does, had forgotten that.

**Review by morgante — APPROVED** [review]
Assuming tests pass, LGTM.

**Comment by ocervell** [comment]
Actually reverting the commit below poses problem for upgrading to this version. I'm getting the error:
```
Error: error archiving directory: could not archive missing directory: .terraform/modules/slos/modules/slo/code_af21402db462

  on .terraform/modules/slos.slo_cloud_function.main/main.tf line 44, in data "archive_file" "main":
  44: data "archive_file" "main" {
```
This happens on re-run apply in CI, because the `.terraform` folder is empty. I guess since the data source runs before the `local_file` resource the folder is not yet created when the `archive_file` wants to read it.

**Comment by ocervell** [comment]
The issue in detail from CI:

* **Run 1:** Everything works as expected, temporary directories to hold GCF code files are created in `.terraform`
* **Run 2:** Data archive_file in terraform-google-event-function runs, but the directories are missing, and we're getting the error `Error: error archiving directory: could not archive missing directory`.
* **Fix:** Remove dependency chain for `terraform-google-scheduled-function` > `terraform-google-event-function` and add native resource code (the following works):

```hcl
resource "random_uuid" "this" {
  keepers = {
    for filename in fileset(local.function_source_directory, "**/*"):
    filename => filemd5("${local.function_source_directory}/${filename}")
  }
}

data "archive_file" "gcf_code" {
  output_path = "${path.root}/.terraform/code-${random_uuid.this.result}.zip"
  type        = "zip"
  dynamic "source" {
    for_each = local.files
    content {
      content = source.value["content"]
      filename = source.value["filename"]
    }
  }
}

resource "google_storage_bucket_object" "archive" {
  name    = "gcf/code-${random_uuid.this.result}.zip"
  bucket  = local.slo_bucket_name
  source  = data.archive_file.gcf_code.output_path
}

resource "google_pubsub_topic" "scheduler_topic" {
  project = var.project_id
  name    = "scheduler-topic-${random_uuid.this.result}"
  message_storage_policy {
    allowed_persistence_regions = [
      var.region
    ]
  }
}

resource "google_cloud_scheduler_job" "scheduler" {
  project  = var.project_id 
  region   = var.region
  name     = "scheduler-job-${random_uuid.this.result}"
  schedule = var.schedule
  pubsub_target {
    topic_name = google_pubsub_topic.scheduler_topic.id
    data       = base64encode("test")
  }
}

resource "google_cloudfunctions_function" "function" {
  project               = var.project_id
  region                = var.region
  name                  = local.full_name
  description           = var.config.slo_description
  runtime               = "python37"
  available_memory_mb   = 128
  source_archive_bucket = local.slo_bucket_name
  source_archive_object = google_storage_bucket_object.archive.name
  entry_point           = "main"
  labels                = var.labels
  event_trigger {
    event_type = "google.pubsub.topic.publish"
    resource   = google_pubsub_topic.scheduler_topic.name
    failure_policy {
      retry = false
    }
  }
  service_account_email = local.sa_email
  environment_variables = var.environment_variables
  vpc_connector = var.vpc_connector
  vpc_connector_egress_settings = var.vpc_connector_egress_settings
}
```

Opened issues:
* Terraform: https://github.com/hashicorp/terraform/issues/26564.
* terraform-google-event-function: https://github.com/terraform-google-modules/terraform-google-event-function/issues/52

@morgante open to suggestions here.

**Pull request closed without merging**